scripts/dotnet-cli-build/Utils/Crossgen.cs

```diff
@@ -42,9 +59,59 @@ public static string GetCrossgenPathForVersion(string coreClrVersion)
             return Path.Combine(
                 Dirs.NuGetPackages,
                 packageId,
-                coreClrVersion,
+                _coreClrVersion,
                 "tools",
                 $"crossgen{Constants.ExeSuffix}");
         }
+
+        public void CrossgenDirectory(BuildTargetContext c, string pathToAssemblies)
+        {
+            // Check if we need to skip crossgen
+            if (string.Equals(Environment.GetEnvironmentVariable("DISABLE_CROSSGEN"), "1"))
+            {
+                c.Warn("Skipping crossgen for because DISABLE_CROSSGEN is set to 1");
+                return;
+            }
+
+            foreach (var file in Directory.GetFiles(pathToAssemblies))
+            {
+                string fileName = Path.GetFileName(file);
+
+                if (fileName == "mscorlib.dll" || fileName == "mscorlib.ni.dll" || !PEUtils.HasMetadata(file))
+                {
+                    continue;
+                }
+
+                string tempPathName = Path.ChangeExtension(file, "readytorun");
+                string sharedFxPath = c.BuildContext.Get<string>("SharedFrameworkPath");
+
+                // HACK
+                // The input directory can be a portable FAT app (example the CLI itself).
+                // In that case there can be RID specific managed dependencies which are not right next to the app binary (example System.Diagnostics.TraceSource).
+                // We need those dependencies during crossgen. For now we just pass all subdirectories of the input directory as input to crossgen.
+                // The right fix -
+                // If the assembly has deps.json then parse the json file to get all the dependencies. Pass these dependencies as input to crossgen.
+                // else pass the current directory of assembly as input to crossgen.
+                var addtionalPaths = Directory.GetDirectories(pathToAssemblies, "*", SearchOption.AllDirectories).ToList();
+                var paths = new List<string>() { sharedFxPath, pathToAssemblies };
+                paths.AddRange(addtionalPaths);
+                var platformAssembliesPaths = string.Join(";", paths.Distinct());
+
+                IList<string> crossgenArgs = new List<string> {
+                    "-readytorun", "-in", file, "-out", tempPathName,
+                    "-platform_assemblies_paths", platformAssembliesPaths
+                };
+
+                var env = new Dictionary<string, string>()
```

Comment:
You could declare this dictionary outside of the loop, since it stays the same for all iterations.